### PR TITLE
Lock wait/notify for scripts

### PR DESCRIPTION
**Pull request by jacobythwaites**
This merge request provides a capability for sse_core transactions to interact through a simple wait/notify mechanism.

See the unit tests in `expr_lock.erl` and the sample `ct_test/svc_expr_SUITE_data/LockTest.xml` integration.

An example use case is a first transaction using Firebase to wake up a smartphone and using `expr_lock:wait/2` in its script to block until the device connects.

 Upon connecting, the device invokes a second, notify transaction which uses `expr_lock:notify/2` in its script to indicate the first transaction can proceed.

### CODE REVIEW

File: apps/svc_expr/ct_test/svc_expr_SUITE_data/LockTest.xml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  {error, Reason} ->
+    put("error", Reason),
+    "Error"
+end.
+    </prop>
+    <reply name='Ok' fields='payload'/>
+    <reply name='Error' fields='error'/>
+  </request>
+
+  <consume name='Notify' service='Expr' clients='Rest' fields='NOTIFY lockId payload'>
```
Comment (miklosduma) [inline apps/svc_expr/ct_test/svc_expr_SUITE_data/LockTest.xml:46]: This is somewhat confusing. The consume is called notify and it has a client attribute? I know notify is a term generally used in threading but in the case of SSE it gets somewhat overloaded.

Comment (miklosduma) [inline apps/svc_expr/ct_test/svc_expr_SUITE_data/LockTest.xml:46]: This is somewhat confusing. The consume is called notify and it has a client attribute? I know notify is a term generally used in threading but in the case of SSE it gets somewhat overloaded.

File: apps/svc_expr/src/script_modules/expr_lock.erl [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+%%% are designed to be called from within svc_expr scripts.
+%%%
+%%% This gen_server is designed to be started under the svc_expr supervisor,
+%%% and it provides the global scripting lock service for all nodes using
+%%% a globally registered process name.
+%%%
+%%% Example script in request/reply that blocks until notified:
+%%% ```
+%%% Prework = fun(LockId) ->
+%%%  % Do something that communicates the lockId
```
Comment (miklosduma) [inline apps/svc_expr/src/script_modules/expr_lock.erl:26]: Mention in code comments that prework must return ok

File: apps/svc_expr/src/script_modules/expr_lock.erl [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+%% Returns the current state record.
+handle_call(state, _From, State) ->
+  {reply, State, State}.
+
+%% Removes the waiting record using the lock id, constructs a notify
+%% message with payload and sends it to the waiting lock process.
+%% The lock process sends the reply to the client.
+handle_cast({notify, LockId, Payload}, State) ->
+  #?STATE{
+    waitlist = Waitlist
```
Comment (miklosduma) [inline apps/svc_expr/src/script_modules/expr_lock.erl:267]: Might be just my lacking Erlang but does this indeed remove the record from state?

File: apps/svc_expr/ct_test/svc_expr_SUITE.erl [hunk 4]
```diff
@@ -483,3 +484,22 @@ test_svc_expr_close_fun(_Config) ->
...
 
   {comment, "Tested svc_expr close function"}.
+
+%% The expr_lock module provides wait/notify functionality for scripts
+%% in svc_expr.
+test_svc_expr_lock(Config) ->
+  _Step1 =
+    import(Config,
+      "LockTest.xml", ?NODE1, ?ADMIN_USER, "/"),
+
```
[the remark this answers is not in the document]
Comment (jacobythwaites) [reply apps/svc_expr/ct_test/svc_expr_SUITE.erl:494]: Please go ahead :)

File: apps/svc_expr/ct_test/svc_expr_SUITE_data/LockTest.xml (hunk 1 above)
Comment (jacobythwaites) [reply apps/svc_expr/ct_test/svc_expr_SUITE_data/LockTest.xml:46]: Oops. Fixed.

Yes, I'm afraid we have to have that confusion. The wait is notified by a consume operation. Can't be helped :)

File: apps/svc_expr/src/script_modules/expr_lock.erl (hunk 2 above)
Comment (jacobythwaites) [reply apps/svc_expr/src/script_modules/expr_lock.erl:26]: Done.

File: apps/svc_expr/src/script_modules/expr_lock.erl (hunk 3 above)
Comment (jacobythwaites) [reply apps/svc_expr/src/script_modules/expr_lock.erl:267]: No, this doesn't and that's correct. The waiting record is removed upon termination of the waiting process, as caught by handle_info/2 below.

**Pull request merged**